test/python/test_bdd_nexus.py

```diff
@@ -0,0 +1,494 @@
+def test_unpublishing_a_nexus_that_is_not_published():
+    "Unpublishing a nexus that is not published."
+
+
+@scenario("features/nexus.feature", "republishing a nexus with a different protocol")
+def test_fail_republishing_a_nexus_with_a_different_protocol():
+    "Republishing a nexus with a different protocol."
+
+
+@scenario("features/nexus.feature", "republishing a nexus with the same protocol")
+def test_republishing_a_nexus_with_the_same_protocol():
+    "Republishing a nexus with the same protocol."
+
+
+@scenario("features/nexus.feature", "publishing a nexus with a crypto-key")
+def test_publishing_a_nexus_with_a_cryptokey():
+    "Publishing a nexus with a crypto-key."
+
+
+@pytest.fixture(scope="module")
+def remote_instance():
+    yield "ms0"
+
+
+@pytest.fixture(scope="module")
+def nexus_instance():
+    yield "ms1"
+
+
+@pytest.fixture(scope="module")
+def base_instances(remote_instance, nexus_instance):
+    yield [remote_instance, nexus_instance]
+
+
+@pytest.fixture(scope="module")
+def nexus_uuid():
+    yield "86050f0e-6914-4e9c-92b9-1237fd6d17a6"
+
+
+@pytest.fixture(scope="module")
+def local_bdev_uri():
+    yield "bdev:///malloc0"
+
+
+@pytest.fixture(scope="module")
+def base_bdevs(mayastor_mod, base_instances):
+    devices = {}
+    for instance in base_instances:
+        uri = "malloc:///malloc0?size_mb=64&blk_size=4096"
+        name = mayastor_mod[instance].bdev.Create(pb.BdevUri(uri=uri)).name
+        devices[instance] = BaseBdev(name, uri)
+    yield devices
+    for instance, bdev in devices.items():
+        mayastor_mod[instance].bdev.Destroy(pb.BdevUri(uri=bdev.uri))
+
+
+@pytest.fixture(scope="module")
+def shared_remote_bdev_uri(mayastor_mod, base_bdevs, remote_instance):
+    name = base_bdevs[remote_instance].name
+    uri = (
+        mayastor_mod[remote_instance]
+        .bdev.Share(pb.BdevShareRequest(name=name, proto="nvmf"))
+        .uri
+    )
+    yield uri
+    mayastor_mod[remote_instance].bdev.Unshare(pb.BdevShareRequest(name=name))
+
+
+@pytest.fixture(scope="module")
+def file_types():
+    yield ["aio", "uring"]
+
+
+@pytest.fixture(scope="module")
+def local_files(file_types):
+    files = {}
+    for type in file_types:
+        path = f"/tmp/{type}-file.img"
+        uri = f"{type}://{path}?blk_size=4096"
+        subprocess.run(
+            ["sudo", "sh", "-c", f"rm -f '{path}' && truncate -s 64M '{path}'"],
+            check=True,
+        )
+        files[type] = LocalFile(path, uri)
+    yield files
+    for path in [file.path for file in files.values()]:
+        subprocess.run(["sudo", "sh", "-c", f"rm -f '{path}'"], check=True)
+
+
+@pytest.fixture(scope="module")
+def find_nexus(mayastor_mod, nexus_instance):
+    def find(uuid):
+        for nexus in mayastor_mod[nexus_instance].ms.ListNexus(pb.Null()).nexus_list:
+            if nexus.uuid == uuid:
+                return nexus
+        return None
+
+    yield find
```

Comment:
what does this yield do?

Reply:
This is fixture that returns a function as its value. It is used to check if a nexus "currently" exists. We cannot create a simple fixture to return the nexus, as it would only be evaluated once per test (the result is cached and reused everywhere the fixture is referenced), and this is usually not what we want. We could use an ordinary function, but then it could only be called within a fixture and each calling fixture would have to request all of the fixtures whose values the function requires. Doing it this way means the dependent fixtures are only declared once.